internal/persistence/sql/migrations/single_table.go

```diff
@@ -0,0 +1,271 @@
+		Relation:  r.Relation,
+		Object:    r.Object,
+		Namespace: r.Namespace.Name,
+		Subject:   sub,
+	}, nil
+}
+
+func (relationTuples) TableName(ctx context.Context) string {
+	return namespaceTableFromContext(ctx)
+}
+
+func (relationTuple) TableName(ctx context.Context) string {
+	return namespaceTableFromContext(ctx)
+}
+
+func NewToSingleTableMigrator(d dependencies) *toSingleTableMigrator {
+	return &toSingleTableMigrator{
+		d:       d,
+		perPage: 100,
+	}
+}
+
+func (m *toSingleTableMigrator) namespaceMigrationBox(n *namespace.Namespace) (*popx.MigrationBox, error) {
+	c, err := m.d.PopConnectionWithOpts(func(d *pop.ConnectionDetails) {
+		d.Options = map[string]string{
+			"migration_table_name": migrationTableFromNamespace(n),
+		}
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return popx.NewMigrationBox(
+		namespaceMigrations,
+		popx.NewMigrator(c, m.d.Logger(), nil, 0),
+		popx.WithTemplateValues(map[string]interface{}{
+			"tableName": tableFromNamespace(n),
+		}),
+	)
+}
+
+func (m *toSingleTableMigrator) getOldRelationTuples(ctx context.Context, n *namespace.Namespace, page, perPage int) (relationTuples, bool, error) {
+	q := m.d.Persister().Connection(ctx).
+		WithContext(context.WithValue(ctx, namespaceCtxKey, n)).
+		Order("object, relation, subject, commit_time").
+		Paginate(page, perPage)
+
+	var res relationTuples
+	if err := q.All(&res); err != nil {
+		return nil, false, sqlcon.HandleError(err)
+	}
+	for _, r := range res {
+		r.Namespace = n
+	}
+	return res, q.Paginator.Page < q.Paginator.TotalPages, nil
+}
+
+func (m *toSingleTableMigrator) insertOldRelationTuples(ctx context.Context, n *namespace.Namespace, rs ...*relationtuple.InternalRelationTuple) error {
+	for _, r := range rs {
+		if r.Subject == nil {
+			return errors.New("subject is not allowed to be nil")
+		}
+
+		m.d.Logger().WithFields(r.ToLoggerFields()).Trace("creating in legacy database")
+
+		if err := m.d.Persister().Connection(context.WithValue(ctx, namespaceCtxKey, n)).Create(&relationTuple{
+			ID:         "testing only",
+			Object:     r.Object,
+			Relation:   r.Relation,
+			Subject:    r.Subject.String(),
+			CommitTime: time.Now(),
+		}); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (m *toSingleTableMigrator) MigrateNamespace(ctx context.Context, n *namespace.Namespace) error {
+	p, ok := m.d.Persister().(*sql.Persister)
+	if !ok {
+		panic("got unexpected persister")
+	}
+
+	for done, page := false, 1; !done; {
+		if err := p.Transaction(ctx, func(ctx context.Context, _ *pop.Connection) error {
+			rs, hasNext, err := m.getOldRelationTuples(ctx, n, page, m.perPage)
+			if err != nil {
+				return err
+			}
+
+			for _, r := range rs {
+				ri, err := r.toInternal()
+				if err != nil {
+					m.d.Logger().WithField("relation_tuple", r).WithField("hint", "https://github.com/ory/keto/issues/661").WithError(err).Warn("Skipping relation tuple, it seems to be in a broken state. Please recreate it manually.")
```

Comment:
Can this actually happen? How would one recover from this, given that we delete the data after the migration is done?

Reply:
As described in the issue, you can theoretically create a tuple that is serialized for the legacy table into a format that is not de-serializable anymore. The only way to recover is to recreate the tuple once the rest is migrated. But you are right, we should maybe not migrate down in this case.